### ISSUE TITLE
freedos no longer working

### ISSUE THREAD

**Issue by severach** [dosemu2/dosemu2#1419]
**Describe the bug**
A description of what the problem is.

**To Reproduce**
Steps to reproduce the behaviour:

**Attach program/game binaries or provide an URL**
If the problem needs specific DOS files to reproduce, please attach
or provide the download URL.

**Attach the log**
It is located in ~/.dosemu/boot.log
To make the log more useful, you may need to enable some logging flags.
See description of -D option in `man dosemu.bin`.

**A regression?**
If you happened to know this problem didn't
exist on dosemu1 or some earlier versions of
dosemu2, please write.

**Additional info**
Please write here if you did any dosemu2 setup
customizations, like installing any custom DOS
or command.com, or altering the dosemu2
configuration settings. Note that freedos, even
from dosemu-freedos package of dosemu1, counts
as a custom DOS and should be mentioned here.
Also make a note if it is a protected or real-mode
program. If you suspect any particular dosemu2
component (DPMI, PIC, PIT etc) to be responsible
for the problem, please write.

**Comment by stsp** [comment]
Unpack on the relative path should
now be fixed.
As for 512 bytes - do you think its a
regression? I think it was always broken.

**Comment by severach** [comment]
Loss of LASTDRIVE appeared between Jul 27 3360ba8df and Aug 5 0904a5e7b.

The bad lredir error message appeared between Nov 26 023380908 and Nov 26 2048caae0.

I switched to fdpp on Dec 4 8c36cb64d The freedos 512 byte limit happened sometime after that.

**Comment by stsp** [comment]
> Loss of LASTDRIVE appeared between Jul 27 3360ba8 and Aug 5 0904a5e.

LASTDRIVE is not lost, but it can't
be used by lredir any more. You can
check that by adding more drives via
$_hdimage option - that one should
work with LASTDRIVE properly. Please
check if that's the case.

> The bad lredir error message appeared between Nov 26

It only became visible by that time.
But lredir couldn't work with LASTDRIVE
some time before that.

> The freedos 512 byte limit happened sometime after that.

This is very strange.
Could you please get back to that commit
and make sure that everything works fine
there? And if so - could you please bisect
that problem?

**Comment by severach** [comment]
````
+0 ~/.dosemu/drive_c
C /home/chris/c
D /home/chris/d
````

Over the versions `bin gnu dosemu kernel.sys command.com fdppkrnl.sys comcom32.exe` and others move around more than a cat on a hot tin roof. I made a shell script to delete these files from my drives and create the symlinks to wherever the files might be on all the drives that need them. This allows me to use the system default CONFIG.SYS, move all lredir smarts to AUTOEXEC, and discard any unnecessary system drives.

The boot files `fdppauto.bat fdppconf.sys fdconfig.sys` must be symlinked from +0 to C. Sometimes +0 boots and sometimes C boots. D is the original location for DOSEMU so my D must have DOSEMU too. D is the only drive that can reliably provide DOSEMU no matter which FDCONFIG.SYS is booting.

$_hdimage works better than lredir but it seems to have a bug. I fixed my FDPPAUTO.BAT (used by FreeDOS) to detect whether C or D is not properly mounted and lredir them into place. This switches back and forth between 2 BAT files to eliminate the 512 byte cache problem when lredir switches the drive. No need to lredir original drives past D to preserve the previous drives. My drives have all necessary files and folders.

````
$_hdimage = "+0 +1"
````

My AUTOEXEC lredirs over C: and D: to fix this.

````
$_hdimage = "+0 /home/chris/d +1"
````

My AUTOEXEC lredirs over C: to fix this. For this to work `DOSEMU` must exist on my D drive. No way to know whether C is mounted to ~/.dosemu/drive_c is or the system's default C drive. The system's C drive fdconfig.sys references D: and cannot be changed. All is fixed when the device=emufs is loaded. 

````
$_hdimage = "/home/chris/c /home/chris/d +1"
````

This is what I want but it is broken. The D drive is mounted as shown before the DOS boot starts. The C drive is mounted as +0 if it contains valid boot files and the read-only system C drive if not. The C drive as shown does not appear until one of the emufs drivers is loaded. For this to work DOSEMU must be on D: because C:\DOSEMU may not be available even if it is on both on C and +0 if the system decides to mount the system default C drive. The FDPP* boot files must be on C and +0 to boot during the switchover.

This has allowed me to test versions back to Dec 2018. All versions to that point have the 512 bug. I removed all lredir code from FDCONFIG.SYS so the 512 bug but won't bother me any more.

'Tis unfortunate that I can't lredir to random drive letters in FreeDOS so I can assigned functions to particular drive letters. Works fine in FDPP.

````
$_hdimage = "+0 /home/chris/d"
````

This also works. My drives have everything.

**Comment by stsp** [comment]
> Over the versions bin gnu dosemu kernel.sys command.com fdppkrnl.sys comcom32.exe
> and others move around more than a cat on a hot tin roof.

We have `install-freedos` and/or `install-otherdos`
packages to do the right thing. Its not bad that you
keep testing the deprecated dosemu-freedos-1.0-bin.tgz
method, but its not going to be "too stable".

> This switches back and forth between 2 BAT files to eliminate the 512 byte cache problem when lredir switches the drive.

So its applicable to bat files as well?
I suppose its now fixed in fdpp, right?
And perhaps it is only applicable to comcom32
but not to freecom?

> The C drive is mounted as +0 if it contains valid boot files and the read-only system C drive if not.

Try to enable $_set_int_hooks and $_force_int_revect

> This has allowed me to test versions back to Dec 2018. All versions to that point have the 512 bug.

Makes sense, thanks for checking.
Next step: figure out to what DOSes and
command.coms this is applicable.
My suspicion is that it is only applicable
to freedos, and for bat files - only to
comcom32 under freedos (which is an
unusual combination).
I suspect that there is no such problem
under ms/pc DOS or any clones.

> 'Tis unfortunate that I can't lredir to random drive letters in FreeDOS

You can if you use the freedos built by
@andrewbird rather than the one from
dosemu-freedos. We also need to make
sure install-freedos script installs exactly
that build. Which is currently not the case:
```
def download_FREEDOS_FULL_ARCHIVES(freedos_url, destination):
    print('Downloading FreeDOS kernel and FreeCOM...')
    for filename in FREEDOS_ARCHIVES_KERNEL_COMMAND:
        download_file(freedos_url + '/repos/base/' + filename + '.zip', destination)
```

Andrew, could you please update the
download URL in the script with your build?

**Comment by andrewbird** [comment]
> 'Tis unfortunate that I can't lredir to random drive letters in FreeDOS

Yes it's a bug in the released FreeDOS kernels 1.0 - 1.2, but you can grab the current build from git here https://github.com/FDOS/kernel/actions. Click on a successful build and open the artifacts snapshot-xxxxxx.zip file, it has both Watcom and GCC builds (you don't need the map files), then rename your choice to kernel.sys.

FreeCOM is also available in a similar fashion, but has the added complexity of language specific versions in the zip file
https://github.com/FDOS/freecom/actions.

Going forward I'm not sure if the upcoming FreeDOS release has the updated kernel, probably better to ask @PerditionC
 what the plan is.

> Andrew, could you please update the
download URL in the script with your build?

Unfortunately the URL is dynamic, and I'm not sure how long old builds are kept for.
